### PR TITLE
infinite scroll

### LINKED ISSUE #15: pages (for PluginListTab)
plugin manager doesn't support pages atm. this hasn't been that big of a problem, but it's going to become more of a problem as more plugins are created.

e.g., currently, the featured tab omits better loading screens and menu background chooser b/c they're on the 2nd page.

I see that there is some code for pages already (particularly: `m_total`, `m_page`, and `m_pageCount` are set in `HandleResponse`) and I've manged to implement a simple 'keep-requesting-all-pages-and-append-them-one-at-a-time' feature.

@codecat, if you're cool with it, I'm happy to implement support for pages. If so, then I have some questions about UI preferences:

Do you have any prior ideas about the UI/UX around pages?

If not, I have 3 ideas:
1. v simple 'request all the pages' -- this would call the API N times for N pages and would load all the images, too. (I guess this isn't ideal, but maybe the bandwidth load is negligible?)
2. 'infinite scroll' type thing -- when the user scrolls to the bottom, a message is shown and the next page is requested. if there are no more pages then the message says that instead.
3. page tabs -- basically just enumerate the pages in a tab bar at the top (always defaults to page=0 (shown as "Page 1")), and load each page when the respective tab is clicked (results cleared between tab clicks).

Any preference between those 3 or other ideas?

### PR DESCRIPTION
**Pull request by XertroV**
fixes #15

note: while in draft this PR probably contains a bunch of dev code -- i didn't know how else to get it to run alongside the normal plugin manager (but this seems to work). This will be removed for the final PR.

explanation of code bits:

**`m_initialListRendered`**: to avoid immediately triggering a request for the 2nd page, `m_initialListRendered` is false for the first frame. (On that frame `UI::GetScrollMaxY()` returns 0 always. We still want to potentially request more pages when `GetScrollMaxY` is 0, though, because this condition is met with a tall thin window.)

**infinite scroll**: it's pretty crude atm. basically we estimate the row height, and if the user get's close to viewing the last row then we initiate a request. The maths is pretty rough and works *well enough* but occasionally a page can be requested early if the user is very near the bottom (common with page 2 or 3). Suggestions welcome wrt improvements, but it's 80% atm so IDK if it's worth bothering with.

### CODE REVIEW

**Comment by XertroV** [comment]
If there's no changes requested, I'll get to removing that dev code so it can be merged cleanly.

**Comment by XertroV** [comment]
Hmm, there's a bug in the initial code where an extra page is requested -- after a request, new plugins are added but the rows are not drawn / measured correctly, which looks like the user has scrolled to the bottom. [910fbf3](https://github.com/openplanet-nl/plugin-manager/pull/16/commits/910fbf30b3542bb58f3378e5845f22105c4d0005) refactors the code to use a 500ms debounce instead, which seems to work well during local testing.

**Comment by codecat** [comment]
For development stuff I recommend removing `PluginManager` from your game folder entirely (eg. `D:\Games\Trackmania\Openplanet\Plugins\PluginManager`) and then cloning the repo in your user folder. I think that should work fine.

**Comment by XertroV** [comment]
> For development stuff I recommend removing `PluginManager` from your game folder entirely (eg. `D:\Games\Trackmania\Openplanet\Plugins\PluginManager`) and then cloning the repo in your user folder. I think that should work fine.

Works great.

I've cleaned up the code into one nice commit -- the previous code w/ commits is here: https://github.com/XertroV/plugin-manager/tree/infinite-scroll-dev-prev

**Comment by codecat** [comment]
Oh don't worry too much about having multiple commits, I can squash and merge everything into 1 commit. I'll put some inline comments in your code.

File: src/Interface/Tabs/PluginList.as [hunk 1]
```diff
@@ -57,8 +59,12 @@ class PluginListTab : Tab
 	void StartRequest()
 	{
 		Clear();
+		StartRequestForPage(0);
+	}
 
+	void StartRequestForPage(int page) {
 		dictionary params;
+		params['page'] = '' + page;
```
Comment (codecat) [inline src/Interface/Tabs/PluginList.as:67]: Please use `tostring(page)` instead as it is a better syntax.

File: src/Interface/Tabs/PluginList.as [hunk 2]
```diff
@@ -57,8 +59,12 @@ class PluginListTab : Tab
 	void StartRequest()
 	{
 		Clear();
+		StartRequestForPage(0);
+	}
 
+	void StartRequestForPage(int page) {
```
Comment (codecat) [inline src/Interface/Tabs/PluginList.as:65]: `{` should be on a new line here.

File: src/Interface/Tabs/PluginList.as [hunk 3]
```diff
@@ -136,8 +144,9 @@ class PluginListTab : Tab
 	{
 		CheckRequest();
 
-		if (m_request !is null) {
+		if (m_request !is null && m_pageCount == 0) {
 			UI::Text("Loading list..");
+			m_lastPageRequestFinished = Time::Now;
```
Comment (codecat) [inline src/Interface/Tabs/PluginList.as:149]: Why is this here? It seems like it's not the intended purpose considering the variable name. Is there a different solution to accomplish the same thing, in a cleaner and more obvious way?

File: src/Interface/Tabs/PluginList.as [hunk 4]
```diff
@@ -159,7 +168,26 @@ class PluginListTab : Tab
...
 			}
+
+			float rowHeight = Math::Ceil(UI::GetScrollMaxY() / (m_plugins.Length / Setting_PluginsPerRow + 1));
+
+			// draw a message before we check for being near the end
+			bool haveMorePages = m_page + 1 < m_pageCount;
+			if (haveMorePages) {
+				UI::TableNextRow(UI::TableRowFlags::None, rowHeight);
+				UI::TableNextColumn();
+				string infiniteScrollMsg = (m_request is null ? "Scroll to Load" : "Loading") + " Page " + (m_page + 2);
```
Comment (codecat) [inline src/Interface/Tabs/PluginList.as:179]: Make it lowercase `page` instead of `Page`

**Review by codecat — CHANGES_REQUESTED** [review]

File: src/Interface/Tabs/PluginList.as (hunk 3 above)
Comment (XertroV) [reply src/Interface/Tabs/PluginList.as:149]: Some leftover code to ensure that the 500ms debounce doesn't start till the first frame that the table is drawn.
That ends up happening in the expected place, though, too. I removed it and there's no apparent difference.

File: src/Interface/Tabs/PluginList.as (hunk 4 above)
Comment (XertroV) [reply src/Interface/Tabs/PluginList.as:179]: I changed "Load" to "load", too. (Though that message is hard to see without furious scrolling)

**Comment by XertroV** [comment]
bump
(just wanted to remind you that this PR exists -- I think all your comments were addressed.)

**Review by codecat — APPROVED** [review]

**Comment by codecat** [comment]
Thank you 😊
